Web Inspector: Web inspector become unresponsive after opening online photos

rcaliman-apple opened the pull request:
#### 9b097d0ecda363b302393bf22b53c771d197995e
<pre>
Web Inspector: Web inspector become unresponsive after opening online photos
<a href="https://bugs.webkit.org/show_bug.cgi?id=254602">https://bugs.webkit.org/show_bug.cgi?id=254602</a>

Reviewed by Patrick Angle and Devin Rousso.

When the inspected main resource is an image, it is treated as a document regardless of its MIME type.

Its resource type is incorrectly assigned on the backend as `InspectorPageAgent::Document`.

See: `InspectorPageAgent::inspectorResourceType(CachedResource::Type type)`
See: `CachedResourceLoader::requestMainResource(CachedResourceRequest&amp;&amp; request)`

The Sources panel on the frontend represents resources of the type
`WI.Resource.Type.Document` with a `WI.TextResourceContentView`.

The payload of the image resource response is a blob of base64 text.

When inspecting standalone images, Web Inspector passes this giant single-line blob of text
to a `WI.SourceCodeTextEditor` where it runs expensive operations on the main thread
and blocks it for long periods of time it if there&apos;s a lot of content.

This patch modifies `InspectorPageAgent` on the backend to return
the appropriate resource type for images, `InspectorPageAgent::ImageResource`,
even when they are the main resource.

This way, the frontend can represent images with the appropriate view: `WI.ImageResourceContentView`.

* LayoutTests/http/tests/inspector/network/resource-mime-type-expected.txt:
* LayoutTests/http/tests/inspector/network/resource-mime-type.html:

Update test for resource type expectations when the document is an image.

* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::inspectorResourceType):

Canonical link: <a href="https://commits.webkit.org/262409@main">https://commits.webkit.org/262409@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22114d71cdca5adb38cf4054cd1623a9fb984d25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1483 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1533 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2376 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1278 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1407 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1466 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1330 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1317 "Failed to checkout and rebase branch from PR 12063") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2226 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1341 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1300 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1263 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1325 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2410 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1378 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1226 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1307 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/368 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1420 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
```diff
@@ -152,11 +152,11 @@ WI.ResourceClusterContentView = class ResourceClusterContentView extends WI.Clus
 
-        this._responseContentView = this._contentViewForResourceType(this._resource.type);
+        let typeFromMIMEType = WI.Resource.typeFromMIMEType(this._resource.mimeType);
+        this._responseContentView = this._contentViewForResourceType(this._resource.type, typeFromMIMEType);
```
this is kinda odd IMO :/

we're trying to find a `WI.ResourceContentView` that matches the `this._resource.type`, but also providing the `typeFromMIMEType` in case we want to use it as an override in other scenarios?

i feel like it'd be more clear/clean to maybe do the MIME attempt first (line 160) and then fall back to the resource type

alternatively, maybe we should adjust `WI.Resource.resolveType` to do this override instead?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
```diff
@@ -249,9 +249,14 @@ WI.ResourceClusterContentView = class ResourceClusterContentView extends WI.Clus
     {
         switch (type) {
+        // When the main resource is an image, it identifies as a document.
+        case WI.Resource.Type.Document:
```
this case exists twice.  please remove the other one

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
```diff
@@ -249,9 +249,14 @@ WI.ResourceClusterContentView = class ResourceClusterContentView extends WI.Clus
+        case WI.Resource.Type.Document:
+            if (typeFromMIMEType === WI.Resource.Type.Image)
+                return new WI.ImageResourceContentView(this._resource);
+
```
please add a `// fallthrough` here so that it's clear it's intentional

dcrousso reviewed (commented):
it seems like this is an issue on the backend, no?  or do we expect/require that the main resource is always a `WI.Resource.Type.Document`?  in that case maybe we should add a `WI.Resource.Type.ImageDocument`?

additionally, does this PR also work for image-able things like SVG?

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -284,6 +284,9 @@ InspectorPageAgent::ResourceType InspectorPageAgent::inspectorResourceType(Cache
 
 InspectorPageAgent::ResourceType InspectorPageAgent::inspectorResourceType(const CachedResource& cachedResource)
 {
+    if (cachedResource.type() == CachedResource::Type::MainResource && MIMETypeRegistry::isSupportedImageMIMEType(cachedResource.mimeType()))
```
are there any other things we could/should do like this?

e.g. if you're directly loading a `.js` resource should we have it be a `InspectorPageAgent::ScriptResource`?

patrickangle replied:
Razvan and I chatted about this this morning when discussing this approach - it would likely be technically correct to do so, but it is also less important to do so since the fallback is to show text form of content anyways. This specific fix is targeting a known bug - a follow up to expand upon this fix wouldn’t hurt though.

dcrousso reviewed (approved):
any way we could add a test for this?  maybe load a page, open Web Inspector, navigate to an image, and then check stuff inside Web Inspector?

rcaliman-apple committed:
Web Inspector: Web inspector become unresponsive after opening online photos
https://bugs.webkit.org/show_bug.cgi?id=254602

Reviewed by Patrick Angle and Devin Rousso.

When the inspected main resource is an image, it is treated as a document regardless of its MIME type.

Its resource type is incorrectly assigned on the backend as `InspectorPageAgent::Document`.

See: `InspectorPageAgent::inspectorResourceType(CachedResource::Type type)`
See: `CachedResourceLoader::requestMainResource(CachedResourceRequest&& request)`

The Sources panel on the frontend represents resources of the type
`WI.Resource.Type.Document` with a `WI.TextResourceContentView`.

The payload of the image resource response is a blob of base64 text.

When inspecting standalone images, Web Inspector passes this giant single-line blob of text
to a `WI.SourceCodeTextEditor` where it runs expensive operations on the main thread
and blocks it for long periods of time it if there's a lot of content.

This patch modifies `InspectorPageAgent` on the backend to return
the appropriate resource type for images, `InspectorPageAgent::ImageResource`,
even when they are the main resource.

This way, the frontend can represent images with the appropriate view: `WI.ImageResourceContentView`.

* LayoutTests/http/tests/inspector/network/resource-mime-type-expected.txt:
* LayoutTests/http/tests/inspector/network/resource-mime-type.html:

Update test for resource type expectations when the document is an image.

* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::inspectorResourceType):

Canonical link: https://commits.webkit.org/262409@main
